File: go/bittorrent/peerwire.go
// SPDX-License-Identifier: MIT

package bittorrent

import (
	"encoding/binary"
	"fmt"
)

const protocolString = "BitTorrent protocol"

// Handshake is the 68-byte BitTorrent peer-wire handshake (BEP-3):
// pstrlen(1)=19 · "BitTorrent protocol"(19) · reserved(8) · info_hash(20) · peer_id(20).
type Handshake struct {
	Reserved [8]byte
	InfoHash [20]byte
	PeerID   [20]byte
}

// DefaultReserved advertises the extension protocol (BEP-10) and DHT (BEP-5).
func DefaultReserved() [8]byte {
	var r [8]byte
	r[5] |= 0x10 // extension protocol
	r[7] |= 0x01 // DHT
	return r
}

// ToBytes serializes the 68-byte handshake.
func (h Handshake) ToBytes() []byte {
	buf := make([]byte, 68)
	buf[0] = 19
	copy(buf[1:20], protocolString)
	copy(buf[20:28], h.Reserved[:])
	copy(buf[28:48], h.InfoHash[:])
	copy(buf[48:68], h.PeerID[:])
	return buf
}

// ParseHandshake parses a 68-byte handshake.
func ParseHandshake(data []byte) (Handshake, error) {
	var h Handshake
	if len(data) < 68 {
		return h, fmt.Errorf("handshake is %d bytes, need 68", len(data))
	}
	if data[0] != 19 {
		return h, fmt.Errorf("handshake pstrlen is %d, want 19", data[0])
	}
	if string(data[1:20]) != protocolString {
		return h, fmt.Errorf("handshake protocol string mismatch")
	}
	copy(h.Reserved[:], data[20:28])
	copy(h.InfoHash[:], data[28:48])
	copy(h.PeerID[:], data[48:68])
	return h, nil
}

// SupportsExtended reports whether the reserved bits advertise BEP-10.
func (h Handshake) SupportsExtended() bool { return h.Reserved[5]&0x10 != 0 }

// SupportsDht reports whether the reserved bits advertise BEP-5.
func (h Handshake) SupportsDht() bool { return h.Reserved[7]&0x01 != 0 }

// MessageType is a BEP-3 peer-wire message id (plus 20 = BEP-10 extended).
type MessageType byte

const (
	MsgChoke         MessageType = 0
	MsgUnchoke       MessageType = 1
	MsgInterested    MessageType = 2
	MsgNotInterested MessageType = 3
	MsgHave          MessageType = 4
	MsgBitfield      MessageType = 5
	MsgRequest       MessageType = 6
	MsgPiece         MessageType = 7
	MsgCancel        MessageType = 8
	MsgPort          MessageType = 9
	MsgExtended      MessageType = 20
)

// PeerMessage is a peer-wire message. A keep-alive has HasID=false (zero-length frame).
type PeerMessage struct {
	HasID   bool
	ID      MessageType
	Payload []byte
}

// KeepAlive is the zero-length keep-alive message.
func KeepAlive() PeerMessage { return PeerMessage{} }

// NewMessage builds a message with an id and payload.
func NewMessage(id MessageType, payload []byte) PeerMessage {
	return PeerMessage{HasID: true, ID: id, Payload: payload}
}

// Simple factories.
func Choke() PeerMessage         { return NewMessage(MsgChoke, nil) }
func Unchoke() PeerMessage       { return NewMessage(MsgUnchoke, nil) }
func Interested() PeerMessage    { return NewMessage(MsgInterested, nil) }
func NotInterested() PeerMessage { return NewMessage(MsgNotInterested, nil) }

func Have(pieceIndex uint32) PeerMessage {
	p := make([]byte, 4)
	binary.BigEndian.PutUint32(p, pieceIndex)
	return NewMessage(MsgHave, p)
}

func BitfieldMsg(bits []byte) PeerMessage { return NewMessage(MsgBitfield, bits) }

func Request(index, begin, length uint32) PeerMessage {
	p := make([]byte, 12)
	binary.BigEndian.PutUint32(p[0:4], index)
	binary.BigEndian.PutUint32(p[4:8], begin)
	binary.BigEndian.PutUint32(p[8:12], length)
	return NewMessage(MsgRequest, p)
}

func Cancel(index, begin, length uint32) PeerMessage {
	m := Request(index, begin, length)
	m.ID = MsgCancel
	return m
}

func Piece(index, begin uint32, block []byte) PeerMessage {
	p := make([]byte, 8+len(block))
	binary.BigEndian.PutUint32(p[0:4], index)
	binary.BigEndian.PutUint32(p[4:8], begin)
	copy(p[8:], block)
	return NewMessage(MsgPiece, p)
}

func Port(port uint16) PeerMessage {
	p := make([]byte, 2)
	binary.BigEndian.PutUint16(p, port)
	return NewMessage(MsgPort, p)
}

func Extended(subID byte, body []byte) PeerMessage {
	p := make([]byte, 1+len(body))
	p[0] = subID
	copy(p[1:], body)
	return NewMessage(MsgExtended, p)
}

// ToBytes serializes the message with its 4-byte big-endian length prefix.
func (m PeerMessage) ToBytes() []byte {
	if !m.HasID {
		return []byte{0, 0, 0, 0} // keep-alive
	}
	length := 1 + len(m.Payload)
	buf := make([]byte, 4+length)
	binary.BigEndian.PutUint32(buf[0:4], uint32(length))
	buf[4] = byte(m.ID)
	copy(buf[5:], m.Payload)
	return buf
}

// ParseBody parses a message body (id + payload, no length prefix). Empty = keep-alive.
func ParseBody(body []byte) (PeerMessage, error) {
	if len(body) == 0 {
		return KeepAlive(), nil
	}
	return NewMessage(MessageType(body[0]), append([]byte(nil), body[1:]...)), nil
}

// ParseFrame parses a full length-prefixed frame, returning the message and bytes consumed.
func ParseFrame(data []byte) (PeerMessage, int, error) {
	if len(data) < 4 {
		return PeerMessage{}, 0, fmt.Errorf("frame shorter than 4-byte length prefix")
	}
	length := binary.BigEndian.Uint32(data[0:4])
	if int(length)+4 > len(data) {
		return PeerMessage{}, 0, fmt.Errorf("frame length %d exceeds available %d", length, len(data)-4)
	}
	msg, err := ParseBody(data[4 : 4+int(length)])
	if err != nil {
		return PeerMessage{}, 0, err
	}
	return msg, 4 + int(length), nil
}

// HavePieceIndex decodes a Have payload.
func (m PeerMessage) HavePieceIndex() (uint32, error) {
	if m.ID != MsgHave || len(m.Payload) != 4 {
		return 0, fmt.Errorf("not a valid have message")
	}
	return binary.BigEndian.Uint32(m.Payload), nil
}

// BlockRef decodes a Request/Cancel payload (index, begin, length).
func (m PeerMessage) BlockRef() (index, begin, length uint32, err error) {
	if (m.ID != MsgRequest && m.ID != MsgCancel) || len(m.Payload) != 12 {
		return 0, 0, 0, fmt.Errorf("not a valid request/cancel message")
	}
	return binary.BigEndian.Uint32(m.Payload[0:4]),
		binary.BigEndian.Uint32(m.Payload[4:8]),
		binary.BigEndian.Uint32(m.Payload[8:12]), nil
}

// PieceBlock decodes a Piece payload (index, begin, block).
func (m PeerMessage) PieceBlock() (index, begin uint32, block []byte, err error) {
	if m.ID != MsgPiece || len(m.Payload) < 8 {
		return 0, 0, nil, fmt.Errorf("not a valid piece message")
	}
	return binary.BigEndian.Uint32(m.Payload[0:4]),
		binary.BigEndian.Uint32(m.Payload[4:8]),
		append([]byte(nil), m.Payload[8:]...), nil
}

// PortValue decodes a Port payload.
func (m PeerMessage) PortValue() (uint16, error) {
	if m.ID != MsgPort || len(m.Payload) != 2 {
		return 0, fmt.Errorf("not a valid port message")
	}
	return binary.BigEndian.Uint16(m.Payload), nil
}

// ── Bitfield (MSB-first: piece 0 is 0x80 of byte 0) ──────────────────────────

type Bitfield struct {
	bits  []byte
	count int
}

// NewBitfield allocates a cleared bitfield for pieceCount pieces.
func NewBitfield(pieceCount int) *Bitfield {
	return &Bitfield{bits: make([]byte, (pieceCount+7)/8), count: pieceCount}
}

// BitfieldFromBytes wraps received bytes for pieceCount pieces.
func BitfieldFromBytes(data []byte, pieceCount int) *Bitfield {
	need := (pieceCount + 7) / 8
	b := make([]byte, need)
	copy(b, data)
	return &Bitfield{bits: b, count: pieceCount}
}

func (b *Bitfield) Count() int { return b.count }

func (b *Bitfield) Get(i int) bool {
	if i < 0 || i >= b.count {
		return false
	}
	return b.bits[i>>3]&(0x80>>(uint(i)&7)) != 0
}

func (b *Bitfield) Set(i int) {
	if i < 0 || i >= b.count {
		return
	}
	b.bits[i>>3] |= 0x80 >> (uint(i) & 7)
}

func (b *Bitfield) PopCount() int {
	n := 0
	for i := 0; i < b.count; i++ {
		if b.Get(i) {
			n++
		}
	}
	return n
}

func (b *Bitfield) HasAll() bool { return b.PopCount() == b.count }

func (b *Bitfield) ToBytes() []byte { return append([]byte(nil), b.bits...) }

File: swift/Sources/AetherNetBitTorrent/PeerWire.swift
// SPDX-License-Identifier: MIT

import Foundation

let protocolString = "BitTorrent protocol"

// ── big-endian byte helpers ──────────────────────────────────────────────────

@inline(__always)
func putUInt32BE(_ v: UInt32, into buf: inout [UInt8], at offset: Int) {
    buf[offset]     = UInt8((v >> 24) & 0xFF)
    buf[offset + 1] = UInt8((v >> 16) & 0xFF)
    buf[offset + 2] = UInt8((v >> 8) & 0xFF)
    buf[offset + 3] = UInt8(v & 0xFF)
}

@inline(__always)
func putUInt16BE(_ v: UInt16, into buf: inout [UInt8], at offset: Int) {
    buf[offset]     = UInt8((v >> 8) & 0xFF)
    buf[offset + 1] = UInt8(v & 0xFF)
}

@inline(__always)
func readUInt32BE(_ buf: [UInt8], at offset: Int) -> UInt32 {
    (UInt32(buf[offset]) << 24) | (UInt32(buf[offset + 1]) << 16)
        | (UInt32(buf[offset + 2]) << 8) | UInt32(buf[offset + 3])
}

@inline(__always)
func readUInt16BE(_ buf: [UInt8], at offset: Int) -> UInt16 {
    (UInt16(buf[offset]) << 8) | UInt16(buf[offset + 1])
}

// ── handshake ────────────────────────────────────────────────────────────────

/// The 68-byte BitTorrent peer-wire handshake (BEP-3):
/// pstrlen(1)=19 · "BitTorrent protocol"(19) · reserved(8) · info_hash(20) · peer_id(20).
public struct Handshake {
    public var reserved: [UInt8]   // 8 bytes
    public var infoHash: [UInt8]   // 20 bytes
    public var peerID: [UInt8]     // 20 bytes

    public init(reserved: [UInt8] = [UInt8](repeating: 0, count: 8),
                infoHash: [UInt8] = [UInt8](repeating: 0, count: 20),
                peerID: [UInt8] = [UInt8](repeating: 0, count: 20)) {
        self.reserved = reserved
        self.infoHash = infoHash
        self.peerID = peerID
    }

    /// Advertises the extension protocol (BEP-10) and DHT (BEP-5).
    public static func defaultReserved() -> [UInt8] {
        var r = [UInt8](repeating: 0, count: 8)
        r[5] |= 0x10  // extension protocol
        r[7] |= 0x01  // DHT
        return r
    }

    /// Serializes the 68-byte handshake.
    public func toBytes() -> [UInt8] {
        var buf = [UInt8](repeating: 0, count: 68)
        buf[0] = 19
        let proto = Array(protocolString.utf8)
        for i in 0..<19 { buf[1 + i] = proto[i] }
        for i in 0..<8 { buf[20 + i] = reserved[i] }
        for i in 0..<20 { buf[28 + i] = infoHash[i] }
        for i in 0..<20 { buf[48 + i] = peerID[i] }
        return buf
    }

    /// Reports whether the reserved bits advertise BEP-10.
    public var supportsExtended: Bool { reserved[5] & 0x10 != 0 }

    /// Reports whether the reserved bits advertise BEP-5.
    public var supportsDht: Bool { reserved[7] & 0x01 != 0 }
}

public enum PeerWireError: Error, Equatable {
    case invalid(String)
}

/// Parses a 68-byte handshake.
public func parseHandshake(_ data: [UInt8]) throws -> Handshake {
    if data.count < 68 {
        throw PeerWireError.invalid("handshake is \(data.count) bytes, need 68")
    }
    if data[0] != 19 {
        throw PeerWireError.invalid("handshake pstrlen is \(data[0]), want 19")
    }
    if Array(data[1..<20]) != Array(protocolString.utf8) {
        throw PeerWireError.invalid("handshake protocol string mismatch")
    }
    return Handshake(
        reserved: Array(data[20..<28]),
        infoHash: Array(data[28..<48]),
        peerID: Array(data[48..<68])
    )
}

// ── messages ─────────────────────────────────────────────────────────────────

/// A BEP-3 peer-wire message id (plus 20 = BEP-10 extended).
public enum MessageType: UInt8 {
    case choke = 0
    case unchoke = 1
    case interested = 2
    case notInterested = 3
    case have = 4
    case bitfield = 5
    case request = 6
    case piece = 7
    case cancel = 8
    case port = 9
    case extended = 20
}

/// A peer-wire message. A keep-alive has `hasID == false` (zero-length frame).
public struct PeerMessage {
    public var hasID: Bool
    public var id: MessageType
    public var payload: [UInt8]

    public init(hasID: Bool = false, id: MessageType = .choke, payload: [UInt8] = []) {
        self.hasID = hasID
        self.id = id
        self.payload = payload
    }

    /// Serializes the message with its 4-byte big-endian length prefix.
    public func toBytes() -> [UInt8] {
        if !hasID {
            return [0, 0, 0, 0]  // keep-alive
        }
        let length = 1 + payload.count
        var buf = [UInt8](repeating: 0, count: 4 + length)
        putUInt32BE(UInt32(length), into: &buf, at: 0)
        buf[4] = id.rawValue
        for i in 0..<payload.count { buf[5 + i] = payload[i] }
        return buf
    }

    /// Decodes a Have payload.
    public func havePieceIndex() throws -> UInt32 {
        if id != .have || payload.count != 4 {
            throw PeerWireError.invalid("not a valid have message")
        }
        return readUInt32BE(payload, at: 0)
    }

    /// Decodes a Request/Cancel payload (index, begin, length).
    public func blockRef() throws -> (index: UInt32, begin: UInt32, length: UInt32) {
        if (id != .request && id != .cancel) || payload.count != 12 {
            throw PeerWireError.invalid("not a valid request/cancel message")
        }
        return (readUInt32BE(payload, at: 0), readUInt32BE(payload, at: 4), readUInt32BE(payload, at: 8))
    }

    /// Decodes a Piece payload (index, begin, block).
    public func pieceBlock() throws -> (index: UInt32, begin: UInt32, block: [UInt8]) {
        if id != .piece || payload.count < 8 {
            throw PeerWireError.invalid("not a valid piece message")
        }
        return (readUInt32BE(payload, at: 0), readUInt32BE(payload, at: 4), Array(payload[8...]))
    }

    /// Decodes a Port payload.
    public func portValue() throws -> UInt16 {
        if id != .port || payload.count != 2 {
            throw PeerWireError.invalid("not a valid port message")
        }
        return readUInt16BE(payload, at: 0)
    }
}

// Message factories (mirror the Go free functions).

public func keepAlive() -> PeerMessage { PeerMessage() }

public func newMessage(_ id: MessageType, _ payload: [UInt8]) -> PeerMessage {
    PeerMessage(hasID: true, id: id, payload: payload)
}

public func choke() -> PeerMessage { newMessage(.choke, []) }
public func unchoke() -> PeerMessage { newMessage(.unchoke, []) }
public func interested() -> PeerMessage { newMessage(.interested, []) }
public func notInterested() -> PeerMessage { newMessage(.notInterested, []) }

public func have(_ pieceIndex: UInt32) -> PeerMessage {
    var p = [UInt8](repeating: 0, count: 4)
    putUInt32BE(pieceIndex, into: &p, at: 0)
    return newMessage(.have, p)
}

public func bitfieldMsg(_ bits: [UInt8]) -> PeerMessage { newMessage(.bitfield, bits) }

public func request(_ index: UInt32, _ begin: UInt32, _ length: UInt32) -> PeerMessage {
    var p = [UInt8](repeating: 0, count: 12)
    putUInt32BE(index, into: &p, at: 0)
    putUInt32BE(begin, into: &p, at: 4)
    putUInt32BE(length, into: &p, at: 8)
    return newMessage(.request, p)
}

public func cancel(_ index: UInt32, _ begin: UInt32, _ length: UInt32) -> PeerMessage {
    var m = request(index, begin, length)
    m.id = .cancel
    return m
}

public func piece(_ index: UInt32, _ begin: UInt32, _ block: [UInt8]) -> PeerMessage {
    var p = [UInt8](repeating: 0, count: 8 + block.count)
    putUInt32BE(index, into: &p, at: 0)
    putUInt32BE(begin, into: &p, at: 4)
    for i in 0..<block.count { p[8 + i] = block[i] }
    return newMessage(.piece, p)
}

public func port(_ portValue: UInt16) -> PeerMessage {
    var p = [UInt8](repeating: 0, count: 2)
    putUInt16BE(portValue, into: &p, at: 0)
    return newMessage(.port, p)
}

public func extended(_ subID: UInt8, _ body: [UInt8]) -> PeerMessage {
    var p = [UInt8](repeating: 0, count: 1 + body.count)
    p[0] = subID
    for i in 0..<body.count { p[1 + i] = body[i] }
    return newMessage(.extended, p)
}

/// Parses a message body (id + payload, no length prefix). Empty = keep-alive.
public func parseBody(_ body: [UInt8]) -> PeerMessage {
    if body.isEmpty {
        return keepAlive()
    }
    let id = MessageType(rawValue: body[0]) ?? .choke
    return newMessage(id, Array(body[1...]))
}

/// Parses a full length-prefixed frame, returning the message and bytes consumed.
public func parseFrame(_ data: [UInt8]) throws -> (PeerMessage, Int) {
    if data.count < 4 {
        throw PeerWireError.invalid("frame shorter than 4-byte length prefix")
    }
    let length = Int(readUInt32BE(data, at: 0))
    if length + 4 > data.count {
        throw PeerWireError.invalid("frame length \(length) exceeds available \(data.count - 4)")
    }
    let msg = parseBody(Array(data[4..<(4 + length)]))
    return (msg, 4 + length)
}

// ── Bitfield (MSB-first: piece 0 is 0x80 of byte 0) ──────────────────────────

public final class Bitfield {
    private var bits: [UInt8]
    public let count: Int

    /// Allocates a cleared bitfield for `pieceCount` pieces.
    public init(pieceCount: Int) {
        self.bits = [UInt8](repeating: 0, count: (pieceCount + 7) / 8)
        self.count = pieceCount
    }

    /// Wraps received bytes for `pieceCount` pieces.
    public init(fromBytes data: [UInt8], pieceCount: Int) {
        let need = (pieceCount + 7) / 8
        var b = [UInt8](repeating: 0, count: need)
        for i in 0..<min(need, data.count) { b[i] = data[i] }
        self.bits = b
        self.count = pieceCount
    }

    public func get(_ i: Int) -> Bool {
        if i < 0 || i >= count { return false }
        return bits[i >> 3] & (0x80 >> UInt8(i & 7)) != 0
    }

    public func set(_ i: Int) {
        if i < 0 || i >= count { return }
        bits[i >> 3] |= 0x80 >> UInt8(i & 7)
    }

    public func popCount() -> Int {
        var n = 0
        for i in 0..<count where get(i) { n += 1 }
        return n
    }

    public func hasAll() -> Bool { popCount() == count }

    public func toBytes() -> [UInt8] { bits }
}
